### ISSUE TITLE
terraform refresh fails for aws_launch_configuration with InvalidAMIID.NotFound

### ISSUE THREAD

**Issue by hashibot** [hashicorp/terraform-provider-aws#673]
_This issue was originally opened by @matt-deboer as hashicorp/terraform#13433. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version
0.9.2

### Affected Resource(s)
- aws_launch_configuration

### Terraform Configuration Files [example]
```hcl
...
## get the ami id from latest packer build
data "aws_ami" "widget" {
  filter {
    name = "name"
    values = ["widget-template"]
  }
  owners = ["self"]
}

resource "aws_launch_configuration" "widget" {
  ...
  image_id = ""${data.aws_ami.widget.id}""
  ...
}

```

Error output:
```
Error refreshing state: 1 error(s) occurred:

* module.aws_agent.aws_launch_configuration.widget: aws_launch_configuration.widget: InvalidAMIID.NotFound: The image id '[ami-xxxxxxxx]' does not exist
	status code: 400, request id: a892c167-7942-4781-8af8-8f62dc57437a
```

### Expected Behavior
`terraform refresh` should be able to succeed, even when the AMI associated with the current aws_launch_configuration has been deleted.
`terraform plan` should show the resources (and affected dependencies) as needing change/rebuild.

### Actual Behavior
`terraform refresh` encounters the error mentioned above

### Steps to Reproduce
1. Construct an terraform configuration, using an `aws_launch_configuration` which pulls it's `image_id` from a datasource, as in the example.
2. `terraform apply` to create the resources
3. delete the AMI (replace with a new/updated version)
4. run `terraform refresh` on the stack

### Important Factoids
Our typical process involves the following steps:
1. Build AMI with packer, and upload to AWS
2. Provision ASGs using an `aws_ami` data source to reference the latest version of the AMI built in packer
3. More recently, we've started cleaning up old AMI's that are no-longer needed, including some of the AMIs produced in step#1, after we have uploaded a newer version of that AMI.
-- we did confirm that there is exactly 1 version of the particular AMI present, but none of the previous versions.
4. We're able to work around this by using `terraform state rm` on the affected `aws_launch_configuration` instances (since they'll be re-created anyway)

**Comment by lukasbernat** [comment]
it would be nice to fix this. it is a pain to fix this once encountered, especially with a large amount of launch configurations.

**Comment by kamilhristov** [comment]
This is still happening in v0.11.1 😞

**Comment by jasonedwardsleadx** [comment]
This is also still happening in v0.11.6.

**Comment by argusua** [comment]
Same for v0.11.7

**Comment by saiya** [comment]
This is still happing in terraform 0.11.6, aws provider 1.14

**Comment by saravanan30erd** [comment]
@argusua @saiya Are you still facing the issue? I couldn't reproduce the issue. If so, please provide the steps in details.

**Comment by saiya** [comment]
I occupied clean VPN and tried to create/modify/delete each `aws_launch_configuration` and AMI. But still failed to reproduce issue...

**Comment by argusua** [comment]
Hi, @saravanan30erd 
I can't reproduce now with:
Terraform v0.11.7
+ provider.aws v1.24.0
+ provider.template v1.0.0

**Comment by saravanan30erd** [comment]
@bflad Looks like the issue is fixed, please verify and close the issue.

**Comment by jasonedwardsleadx** [comment]
This is still an issue for me with:
* provider.aws v1.24.0
* Terraform v0.11.7.

I get:

``` Error: Error refreshing state: 1 error(s) occurred:

* module.asg.aws_launch_configuration.this: 1 error(s) occurred:

* module.asg.aws_launch_configuration.this: aws_launch_configuration.this: InvalidAMIID.NotFound: The image id '[ami-xxxxxxxx]' does not exist
        status code: 400, request id: aea9c0e9-5274-46ea-ac21-703052566614 ```

**Comment by gordonbondon** [comment]
Issue is most probably in this part of code in `resourceAwsLaunchConfigurationRead`:
https://github.com/terraform-providers/terraform-provider-aws/blob/fd9a3f78c2a83c4821b6cf143ee19cc32002c796/aws/resource_aws_launch_configuration.go#L557-L559

Where it tries to read block devices of AMI during resource refresh. This call fails when ami does not exist.

**Comment by FransUrbo** [comment]
Still a case with

Terraform v0.11.8
+ provider.aws v1.33.0
+ provider.terraform v0.1.0

**Comment by FransUrbo** [comment]
Only workaround I found so far:
1) Create a copy of the LC, with a different/new AMI
2) Change the ASG to use that LC instead of the "original" one
3) Delete the "original" LC
4) Run TF - will then create a new, fresh LC and also update the ASG to use that LC instead of the copy/temporary one.

Pain in the behind and cumbersome, especially if you like me have over 40 ASGs spread across several accounts...

**Comment by tiddnet** [comment]
This seems to be still an issue:

`Error: Error refreshing state: 2 error(s) occurred:

* module.WURFLjs-ap-southeast-1.aws_launch_configuration.WURFLjs-go-green: 1 error(s) occurred:

* module.WURFLjs-ap-southeast-1.aws_launch_configuration.WURFLjs-go-green: aws_launch_configuration.WURFLjs-go-green: InvalidAMIID.NotFound: The image id '[ami-xxxxx]' does not exist
	status code: 400, request id: ae8a4a35-6ace-4732-92e1-478a4b03c575
* module.WURFLjs-us-east-2.aws_launch_configuration.WURFLjs-go-green: 1 error(s) occurred:

* module.WURFLjs-us-east-2.aws_launch_configuration.WURFLjs-go-green: aws_launch_configuration.WURFLjs-go-green: InvalidAMIID.NotFound: The image id '[ami-xxxxx]' does not exist
	status code: 400, request id: 7d39284f-50b1-4361-9a9b-2d1b0cf7a0ad


Terraform v0.11.8
+ provider.aws v1.38.0`

**Comment by ekowcharles** [comment]
This issue is still occurring for terraform v0.11.10 aws provider version 1.4 template version 1.0

**Comment by dgreene-r7** [comment]
Still a bug:
```
Terraform v0.11.10
+ provider.aws v1.50.0
```

**Comment by marking** [comment]
anyone have a workaround to this? I tried tainting the resource in the state file. but received the error message that the resource cannot be tainted.

**Comment by dev-head** [comment]
looks like this is still an issue; fingers crossed it's addressed before the issue's two year anniversary 👍 

tried deleting out of state to no avail; looks like an api call is the culprit. going to try a few other ideas.

**Comment by ruke47** [comment]
I ran into this issue; my workaround was to delete the state for the impacted launch configuration:

    terraform state rm module.asg.aws_launch_configuration.widget

When I ran `terraform apply`, it created a new launch configuration, and I had to manually delete the old one.

**Comment by vladimir-shopov-axway** [comment]
The same issue happened to me with Terraform v0.11.14 + provider.aws v2.22.0. Unfortunately I'm unable to reproduce it with a plain Terraform snippet that can be posted here. It appears to be happening under particular circumstances, which I don't know at the moment.

**Comment by maartenschalekamp** [comment]
Also experienced this issue

> Terraform v0.11.14
+ provider.aws v2.32.0

**Comment by Erokos** [comment]
Hi,
experiencing the same issue with
```
Terraform v0.11.13
+ provider.aws v2.34.0
```

**Comment by Xat59** [comment]
Hello guys, 

same behavior with :
```
$ terraform --version
Terraform v0.12.12
```

**Comment by s4ntos** [comment]
Still occuring with :

```
Terraform v0.12.13
+ provider.aws v2.29.0
```

**Comment by conormancone-cimpress** [comment]
Still still occurring with:

```
Terraform v0.12.24
+ provider.aws v2.56.0
+ provider.template v2.1.2
```

**Comment by evandam** [comment]
Seeing same issue still with:

```
Terraform v0.12.24
+ provider.aws v2.61.0
```

Does anyone have updates on this issue?

**Comment by PranshuSharma1** [comment]
I also found the same issue with 
`Terraform v0.12.25`

But I have solved the issue with a trick, 
==> As I created a local which contains the AMI-ID of the AMI image created by the packer.
like this 
```
locals {
  ami_ID = data.aws_ami.<AMI_resource_name>.id
}

```
Just define the local block in the root module or file, just before using the AWS launch template and after the data block of the aws_ami.

And in the AWS launch template in the image_id use the local variable like this:-
```

resource "aws_launch_template" "ALT" {
  name = "NEW-ALT"
  image_id      = local.ami_ID
............
}


```
And this works just great.

**Comment by tyc85** [comment]
@PranshuSharma1 
   How did this approach avoid the issue? In my case this approach still doesn't work. But i am using launch config rather than launch template, which may be the difference.

**Comment by fatalglitch** [comment]
Ran into this issue today, has it really been almost 4 years and this issue isn't resolved?

**Comment by aaron-ballard-530** [comment]
Ran into the issue as well.  Had to remove the resources from the state file and adjust them by hand.

**Comment by thom-vend** [comment]
Still happening, what's the best workaround so far ?

**Comment by bedge** [comment]
Anyone try just deleting the affected launch template from AWS, then re-`apply`ing?
This "works", in that terraform will recreate the entire launch template if it's missing entirely
... at least, on an otherwise "healthy" config.
ie: I haven't tested this on the specific target env that has this actual problem. (I wanted to have a proven solution before I did this on the problem target)

If this is viable, it's a lot simpler than what's been proposed so far.

**Comment by mramitanand** [comment]
> Anyone try just deleting the affected launch template from AWS, then re-`apply`ing? This "works", in that terraform will recreate the entire launch template if it's missing entirely ... at least, on an otherwise "healthy" config. ie: I haven't tested this on the specific target env that has this actual problem. (I wanted to have a proven solution before I did this on the problem target)

I just had the same problem and did exactly what you said - deleted the launch configurations in AWS and that worked great. Thank you!!

**Comment by brittandeyoung** [comment]
Taking a look at this issue now. 

From looking at the code, this issue should already be resolved in the `4.17.0` provider version and greater. 

This code here should allow the terraform refresh to take place even if the `ami_image` is no longer valid:

https://github.com/hashicorp/terraform-provider-aws/blob/3a0df65c8b967caf18ad2816d08d20cf5b467e74/internal/service/autoscaling/launch_configuration.go#LL480C1-L487C3

It looks like this was implemented through this PR: https://github.com/hashicorp/terraform-provider-aws/pull/17695 but the association was not made with the issue. 

Closing out this issues. If any one is able to reproduce this issue on a provider version newer than `4.17.0` . Please comment.
